securesystemslib/signer.py

```diff
@@ -0,0 +1,144 @@
+"""
+<Program Name>
+  signer.py
+
+<Author>
+  Martin Vrachev <mvrachev@vmware.com>
+
+<Started>
+  Januart 27, 2021.
+
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
+  The goal of this module is to provide signing interface supporting multiple
+  signing implementations.
+"""
+
+import abc
+import securesystemslib.keys as sslib_keys
+
+
+
+class Signature:
+  """
+  <Purpose>
+    Storage class containing information about a signature and the keyid
+    uniquely identifying the key used to generate the signature.
+
+  <Attributes>
+    keyid: HEX string used as a unique identifier of the key.
+    signature: HEX string representing the signature.
+  """
+  def __init__(self, keyid, sig):
+    self.keyid = keyid
+    self.signature = sig
+
+
+  @classmethod
+  def from_dict(cls, signature_dict):
+    """
+    <Purpose>
+      Creates a Signature object from its JSON/dict representation.
+
+    <Arguments>
+      signature_dict: A dict containing that should contain valid keyid
+        and signature.
+        Note that the fields should be named "keyid" and "sig" respectively.
+    """
+
+    return cls(signature_dict["keyid"], signature_dict["sig"])
+
+
+  def to_dict(self):
+    """
+    <Purpose>
+      Returns the JSON-serializable dictionary representation of self.
+    """
+    return {
+      "keyid": self.keyid,
+      "sig": self.signature
+      }
+
+
+
+class Signer:
+  """
+  <Purpose>
+    Signer interface created to support multiple signing implementations.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractmethod
+  def sign(payload):
+    """
+    <Purpose>
+      Abstract function used for signing a given payload by the key assigned
+      to the Signer instance.
+
+    <Arguments>
+      payload: bytes to be signed
+
+    <Returns>
+      Returns a "Signature" class instance containing the signature and the
+      the keyid which uniquely identifies the key used for signature generation.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+
+class SSlibSigner(Signer):
+  """
+    <Purpose>
+      Securesystemslib default implementation of the "Signer" interface.
+      With this implementation the following signature schemes are supported:
+
+      'RSASSA-PSS'
```

Comment:
We actually support pkcs1v15 as well. And I think we should use the securesystemslib scheme strings here, as e.g. defined in formats:

```python
 RSA_SCHEME_SCHEMA = SCHEMA.OneOf([
  SCHEMA.RegularExpression(r'rsassa-pss-(md5|sha1|sha224|sha256|sha384|sha512)'),
  SCHEMA.RegularExpression(r'rsa-pkcs1v15-(md5|sha1|sha224|sha256|sha384|sha512)')])
```